core-bundle/src/DataContainer/DataContainerOperationsBuilder.php

```diff
@@ -0,0 +1,270 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\DataContainer;
+
+use Contao\Backend;
+use Contao\CoreBundle\Security\ContaoCorePermissions;
+use Contao\DataContainer;
+use Contao\Image;
+use Contao\Input;
+use Contao\StringUtil;
+use Contao\System;
+use Symfony\Bundle\SecurityBundle\Security;
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+use Twig\Environment;
+
+/**
+ * @internal
+ */
+class DataContainerOperationsBuilder
+{
+    public function __construct(
+        private readonly Environment $twig,
+        private readonly Security $security,
+        private readonly UrlGeneratorInterface $urlGenerator,
+    ) {
+    }
+
+    public function generateButtons(string $table, array $record, DataContainer $dataContainer, callable|null $legacyCallback = null): string|\Stringable
+    {
+        if (!\is_array($GLOBALS['TL_DCA'][$table]['list']['operations'] ?? null)) {
+            return '';
+        }
+
+        $operations = [];
+
+        foreach ($GLOBALS['TL_DCA'][$table]['list']['operations'] as $k => $v) {
+            $v = \is_array($v) ? $v : array($v);
+            $operation = $this->generateOperation($k, $v, $table, $record, $dataContainer, $legacyCallback);
+
+            if ($operation) {
+                $operations[] = $operation;
+            }
+        }
+
+        return $this->createBuilder($operations);
+    }
+
+    public function generateHeaderButtons(string $table, array $record, DataContainer $dataContainer, callable|null $legacyCallback = null): string|\Stringable
+    {
+        if (!\is_array($GLOBALS['TL_DCA'][$table]['list']['operations'] ?? null)) {
+            return '';
+        }
+
+        $operations = [];
+
+        foreach ($GLOBALS['TL_DCA'][$table]['list']['operations'] as $k => $v) {
+            // Edit operation should be shown in header by default
```

Comment:
We should not make this assumption. Adding `'showInHeader' => true` in the default operations listener has the same effect but can be unset if necessary.